samza-api/src/main/java/org/apache/samza/container/grouper/stream/SystemStreamPartitionGrouper.java

```diff
@@ -37,5 +36,9 @@
  * the TaskNames.
  */
 public interface SystemStreamPartitionGrouper {
-  public Map<TaskName, Set<SystemStreamPartition>> group(Set<SystemStreamPartition> ssps);
+  Map<TaskName, Set<SystemStreamPartition>> group(Set<SystemStreamPartition> systemStreamPartitions);
+
+  default Map<TaskName, Set<SystemStreamPartition>> group(Set<SystemStreamPartition> systemStreamPartitions, GrouperContext grouperContext) {
```

Comment:
Is this interface change really necessary. The ```GrouperContext``` can be passed to the interface-implementing class in its constructor as an Option<GrouperContext>. 
If a GrouperContext is specified, the class can take it into consideration on invocations of ```group(ssps)```.

Reply:
Yes, good point. We want to make all the previous metadata of the job execution available to the interface so that it can generate grouping that is partition aware(or in future do other optimizations). 

1. Currently in samza, `SystemStreamPartitionGrouper` is only instantiated through the `SystemStreamPartitionGrouperFactory` using reflection. So, to invoke the appropriate `SystemStreamPartitionGrouper` constructor which takes in the `GrouperContext` optional, the factory has to be aware of the GrouperContext optional. To accomplish this, we need to make a interface change(add a new method) in `SystemStreamPartitionGrouperFactory`. However, the problem then becomes the factory invoking the right constructor and any custom implementations in the future, would come to know that this `GrouperContext` is available by following a trial through the factory which may not be obvious. By adding this to the interface we make the interface contract clear and upfront.
2. I think the grouper context will be present for a job all the time. During the first run of the samza job, it will be empty. I'm not sure of the necessity to make it optional though.

This interface change was approved in SEP-5. If you have strong opinions about this, then lets discuss offline.